On nightly rustc, E0599 emits a massive diagnostic to suggest wrapping the receiver in every combination of `{Box,Pin,Rc,Arc}::new({,&,&mut} expr)`

Given the following code ([playground](https://play.rust-lang.org/?version=beta&mode=debug&edition=2018&gist=d129ca6f7559fe24438efcefc14eec4c)):

```rust
trait Bar {}
impl Bar for String {}

fn main() {
  let s = String::from("hey");
  let x: &dyn Bar = &s;
  x.as_ref();
}
```

The output on `1.53.0-nightly (2021-04-29 478a07df05e3fe8df964)` is:

<details>
<summary>Long diagnostic</summary>

```
error[E0599]: the method `as_ref` exists for reference `&dyn Bar`, but its trait bounds were not satisfied
   --> src/main.rs:7:5
    |
1   | trait Bar {}
    | --------- doesn't satisfy `dyn Bar: AsRef<_>`
...
7   |   x.as_ref();
    |     ^^^^^^ method cannot be called on `&dyn Bar` due to unsatisfied trait bounds
    |
    = note: the following trait bounds were not satisfied:
            `dyn Bar: AsRef<_>`
            which is required by `&dyn Bar: AsRef<_>`
help: consider wrapping the receiver expression with the appropriate type
    |
7   |   Box::new(x).as_ref();
    |   ^^^^^^^^^ ^
help: consider wrapping the receiver expression with the appropriate type
    |
7   |   Pin::new(x).as_ref();
    |   ^^^^^^^^^ ^
help: consider wrapping the receiver expression with the appropriate type
    |
7   |   Arc::new(x).as_ref();
    |   ^^^^^^^^^ ^
help: consider wrapping the receiver expression with the appropriate type
    |
7   |   Rc::new(x).as_ref();
    |   ^^^^^^^^ ^
help: consider wrapping the receiver expression with the appropriate type
    |
7   |   Box::new(&mut x).as_ref();
    |   ^^^^^^^^^^^^^  ^
help: consider wrapping the receiver expression with the appropriate type
    |
7   |   Pin::new(&mut x).as_ref();
    |   ^^^^^^^^^^^^^  ^
help: consider wrapping the receiver expression with the appropriate type
    |
7   |   Arc::new(&mut x).as_ref();
    |   ^^^^^^^^^^^^^  ^
help: consider wrapping the receiver expression with the appropriate type
    |
7   |   Rc::new(&mut x).as_ref();
    |   ^^^^^^^^^^^^  ^
help: consider wrapping the receiver expression with the appropriate type
    |
7   |   Box::new(&x).as_ref();
    |   ^^^^^^^^^^ ^
help: consider wrapping the receiver expression with the appropriate type
    |
7   |   Pin::new(&x).as_ref();
    |   ^^^^^^^^^^ ^
help: consider wrapping the receiver expression with the appropriate type
    |
7   |   Arc::new(&x).as_ref();
    |   ^^^^^^^^^^ ^
help: consider wrapping the receiver expression with the appropriate type
    |
7   |   Rc::new(&x).as_ref();
    |   ^^^^^^^^^ ^
```
</details>

`Box`, `Pin`, `Arc`, and `Rc` are suggested three times each: once for `x`, once for `&x`, and once for `&mut x`. I'm not sure if this is a problem, but this seems incredibly verbose to me.

Note that the output on stable 1.51 and beta 1.52 is different:
```
error[E0599]: the method `as_ref` exists for reference `&dyn Bar`, but its trait bounds were not satisfied
 --> src/main.rs:7:5
  |
1 | trait Bar {}
  | --------- doesn't satisfy `dyn Bar: AsRef<_>`
...
7 |   x.as_ref();
  |     ^^^^^^ method cannot be called on `&dyn Bar` due to unsatisfied trait bounds
  |
  = note: the following trait bounds were not satisfied:
          `dyn Bar: AsRef<_>`
          which is required by `&dyn Bar: AsRef<_>`
  = help: items from traits can only be used if the trait is implemented and in scope
  = note: the following trait defines an item `as_ref`, perhaps you need to implement it:
          candidate #1: `AsRef`
```

--- a/compiler/rustc_typeck/src/check/method/suggest.rs
+++ b/compiler/rustc_typeck/src/check/method/suggest.rs
@@ -579,6 +579,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 }
 
                 let mut restrict_type_params = false;
+                let mut unsatisfied_bounds = false;
                 if !unsatisfied_predicates.is_empty() {
                     let def_span = |def_id| {
                         self.tcx.sess.source_map().guess_head_span(self.tcx.def_span(def_id))
@@ -739,6 +740,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                         err.note(&format!(
                             "the following trait bounds were not satisfied:\n{bound_list}"
                         ));
+                        unsatisfied_bounds = true;
                     }
                 }
 
@@ -752,6 +754,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                         source,
                         out_of_scope_traits,
                         &unsatisfied_predicates,
+                        unsatisfied_bounds,
                     );
                 }
 
@@ -984,9 +987,10 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         source: SelfSource<'tcx>,
         valid_out_of_scope_traits: Vec<DefId>,
         unsatisfied_predicates: &[(ty::Predicate<'tcx>, Option<ty::Predicate<'tcx>>)],
+        unsatisfied_bounds: bool,
     ) {
         let mut alt_rcvr_sugg = false;
-        if let SelfSource::MethodCall(rcvr) = source {
+        if let (SelfSource::MethodCall(rcvr), false) = (source, unsatisfied_bounds) {
             debug!(?span, ?item_name, ?rcvr_ty, ?rcvr);
             let skippable = [
                 self.tcx.lang_items().clone_trait(),

--- a/src/test/ui/suggestions/import-trait-for-method-call.rs
+++ b/src/test/ui/suggestions/import-trait-for-method-call.rs
@@ -6,4 +6,11 @@ fn next_u64() -> u64 {
     h.finish() //~ ERROR no method named `finish` found for struct `DefaultHasher`
 }
 
-fn main() {}
+trait Bar {}
+impl Bar for String {}
+
+fn main() {
+    let s = String::from("hey");
+    let x: &dyn Bar = &s;
+    x.as_ref(); //~ ERROR the method `as_ref` exists for reference `&dyn Bar`, but its trait bounds
+}

--- a/src/test/ui/suggestions/import-trait-for-method-call.stderr
+++ b/src/test/ui/suggestions/import-trait-for-method-call.stderr
@@ -15,6 +15,22 @@ help: the following trait is implemented but not in scope; perhaps add a `use` f
 LL | use std::hash::Hasher;
    |
 
-error: aborting due to previous error
+error[E0599]: the method `as_ref` exists for reference `&dyn Bar`, but its trait bounds were not satisfied
+  --> $DIR/import-trait-for-method-call.rs:15:7
+   |
+LL | trait Bar {}
+   | --------- doesn't satisfy `dyn Bar: AsRef<_>`
+...
+LL |     x.as_ref();
+   |       ^^^^^^ method cannot be called on `&dyn Bar` due to unsatisfied trait bounds
+   |
+   = note: the following trait bounds were not satisfied:
+           `dyn Bar: AsRef<_>`
+           which is required by `&dyn Bar: AsRef<_>`
+   = help: items from traits can only be used if the trait is implemented and in scope
+   = note: the following trait defines an item `as_ref`, perhaps you need to implement it:
+           candidate #1: `AsRef`
+
+error: aborting due to 2 previous errors
 
 For more information about this error, try `rustc --explain E0599`.

